### PR TITLE
Overhaul the API

### PR DESCRIPTION
**Pull request by scrogson**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,10 +14,9 @@ Current implemented strategies:
 #### Authorization Code Flow (AuthCode Strategy)
 
 ```elixir
-alias OAuth2.Strategy.AuthCode
-
-# Initialize the strategy with your client_id, client_secret, and site.
-strategy = AuthCode.new([
+# Initialize a client with your client_id, client_secret, and site.
```
Comment (knewter) [inline README.md]: "initialize ... with strategy, client_id...."

File: README.md [hunk 2]
```diff
@@ -25,11 +24,18 @@ strategy = AuthCode.new([
 ])
 
 # Generate the authorization URL and redirect the user to the provider.
-AuthCode.authorize_url(strategy, %{redirect_uri: strategy.redirect_uri})
+OAuth2.Client.authorize_url(client)
```
Comment (knewter) [inline README.md]: obviously much better :D :+1:

File: README.md [hunk 3]
```diff
@@ -25,11 +24,18 @@ strategy = AuthCode.new([
...
 # => "https://auth.example.com/oauth/authorize?client_id=client_id&redirect_uri=https%3A%2F%2Fexample.com%2Fauth%2Fcallback&response_type=code"
 
 # Use the authorization code returned from the provider to obtain an access token.
-token = AuthCode.get_token!(strategy, "someauthcode", %{redirect_uri: strategy.redirect_uri})
+token = OAuth2.Client.get_token!(client, code: "someauthcode")
```
Comment (knewter) [inline README.md]: I see your wisdom, good sir!

File: lib/oauth2.ex [hunk 4]
```diff
@@ -1,58 +1,65 @@
...
+      token = OAuth2.Client.get_token!(client, code: "someauthcode")
+
+  You can also use `OAuth2.Client.put_param/3` to update the client's `params` field.
+
+  Example:
+
+      token =
+        client
+        |> OAuth2.Client.put_param(:code, "someauthcode")
+        |> OAuth2.Client.get_token!()
```
Comment (knewter) [inline lib/oauth2.ex]: love it

File: lib/oauth2/access_token.ex [hunk 5]
```diff
@@ -1,40 +1,73 @@
...
       expires_at:    (std["expires_in"] || other["expires"]) |> expires_at(),
-      token_type:    response["token_type"] |> normalize_token_type(),
+      token_type:    std["token_type"] |> normalize_token_type(),
       other_params:  other,
-      strategy:      strategy]
+      client:        client]
   end
 
+  @doc """
+  Makes a `GET` request the given URL using the AccessToken.
```
Comment (knewter) [inline lib/oauth2/access_token.ex]: "request to the given URL..."

File: lib/oauth2/access_token.ex [hunk 6]
```diff
@@ -1,40 +1,73 @@
...
+  """
   def get(token, url, headers \\ [], opts \\ []) do
     case Request.get(process_url(token, url), req_headers(token, headers), opts) do
       {:ok, response} -> {:ok, response.body}
       {:error, reason} -> {:error, %Error{reason: reason}}
     end
   end
+
+  @doc """
+  Makes a `GET` request the given URL using the AccessToken.
```
Comment (knewter) [inline lib/oauth2/access_token.ex]: "to the given URL"

File: lib/oauth2/client.ex [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    %{client | headers: List.keystore(headers, key, 0, {key, value})}
+  end
+
+  @doc """
+  Set multiple headers in the client in one call.
+  """
+  @spec put_headers(t, list) :: t
+  def put_headers(%Client{} = client, []), do: client
+  def put_headers(%Client{} = client, [{k,v}|rest]) do
+    client |> put_header(k,v) |> put_headers(rest)
```
Comment (knewter) [inline lib/oauth2/client.ex]: really enjoyed this line

**Comment by knewter** [comment]
Really like all these changes.  They're obviously well thought out.  Almost like you've been using this lib and found some warts :)

**Comment by niftyn8** [comment]
:) :+1: Good changes.

**Pull request merged**